### PR TITLE
Optional inputs, schema validation and docs prettification

### PR DESCRIPTION
**Pull request by dzhulgakov**
Addresses #151 

We still have flexible stuff around NumInputs with arbitrary c function. It seems it's not used that much so we can probably clean up it separately. For now optional inputs are introduced as a separate concept.

### CODE REVIEW

File: onnx/defs/schema.h [hunk 1]
```diff
@@ -258,6 +272,27 @@ class OpSchema {
 };
 
 /**
+ * Internal class used in schema declaration
+ */
+class OpSchemaHolder {
+ public:
+  OpSchemaHolder(OpSchema& schema) : schema_(&schema) {
```
Comment (bddppq) [inline onnx/defs/schema.h]: nit: there is no need to store `schema_`

**Comment by bddppq** [comment]
Are we also going to add optional output?

File: onnx/defs/schema.h (hunk 1 above)
Comment (dzhulgakov) [reply onnx/defs/schema.h]: I kept it just in case somebody would want to access schemas through global variables (probably not)

**Comment by dzhulgakov** [comment]
For optional output it's less critical - we can always map to a random name and that way there's less to handle in backend. With inputs the situation is that sometimes we need to independently enable/disable several inputs and it's hard to express otherwise.

**Comment by yuanbyu** [comment]
For outputs, I think we are good: If an output is not used in the graph, it is optional. A runtime may choose not to generate an output if it can detect the output is not used.

**Pull request merged**